<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
        "http://www.w3.org/TR/html4/loose.dtd">
<html>
    
    <TITLE>README - Examples</TITLE>
    <head>
        <style type="text/css">
    
            .tab { margin-left: 40px; }
            
            .code { color: Teal ; 
                    font-size: 100%; 
                    font-style: italic; 
                    line-height: 1.5;
                    padding-left: 20px;
                    border-style: solid;
                    margin-left: 20px;
                    margin-right: 200px;
            }
            
            .s { color: Teal ; 
                    font-size: 100%; 
                    font-style: italic; 
                    line-height: 1.5;
                    padding-left: 20px;
                    border-style: solid;
                    margin-left: 40px;
                    margin-right: 200px;
            }
            
            
        </style>
    </head>
    
    <body style='background-color:#F2F2F2; font-size:14px'>
                
        <h1>Application Example</h1> 

        <h2>Scan the hg19 human genome with the chen10 CTCF Position Weight Matrix (PWM)</h2>

        <h3> Important remarks: </h3>
        <p class="tab">
          The PWMScan programs rely on a few rules to make output conversion to BED format easier, namely FASTA header and file naming conventions.
        <p class="tab">

        
        <h4> 
            <p class="tab">
            FASTA headers  
        </h4>
        
        <p class="tab">
        All genome sequence files have a FASTA header that is formatted as follows:
        
        <pre>
            <p class="tab">
            >chr|NC_000001|NC_000001.10 <i>some text</i>
        </pre>
        
        <p class="tab">
        The <i><font color=blue>sequence identifier</i></font> includes three words concatenated via a UNIX pipe '|': the word 'chr' followed by the NCBI RefSeq partial and full accession identifiers. 
        The accession records indicate sequence identity. The sequence identifier is parsed by most of the PWMScan programs and utilities.
        <p class="tab">
        For each genome assembly, we provide a table for NCBI RefSeq to chromosome number mapping (<i><font color=blue>chr_NC_gi</i></font>) as well as a table for FASTA sequence identifier to chromosome 
        number conversion (<i><font color=blue>chr_hdr</i></font>). The <i><font color=blue>chr_NC_gi</i></font> file has been downloaded from NCBI whereas the <i><font color=blue>chr_hdr</i></font> 
        file has been generated by us. By consequence, the <i><font color=blue>chr_hdr</i></font> table can be changed to adapt to other FASTA header format conventions.
        <pre>
            <p class="tab">
            <i><font color=firebrick>chr_NC_gi</i></font>
            #Chr    Accession.ver   gi
            1       NC_000001.10    224384768
            2       NC_000002.11    224384767
            3       NC_000003.11    224384766
            4       NC_000004.11    224384765
            5       NC_000005.9     224384764
            6       NC_000006.11    224384763
            7       NC_000007.13    224384762
            8       NC_000008.10    224384761
            9       NC_000009.11    224384760
            10      NC_000010.10    224384759
            11      NC_000011.9     224384758
            12      NC_000012.11    224384757
            13      NC_000013.10    224384756
            14      NC_000014.8     224384755
            15      NC_000015.9     224384754
            16      NC_000016.9     224384753
            17      NC_000017.10    224384752
            18      NC_000018.9     224384751
            19      NC_000019.9     224384750
            20      NC_000020.10    224384749
            21      NC_000021.8     224384748
            22      NC_000022.10    224384747
            X       NC_000023.10    224384746
            Y       NC_000024.9     224384745
            M       NC_012920.1     251831106
        </pre>
        <pre>
            <p class="tab">
            <i><font color=firebrick>chr_hdr</i></font>
            #Chr    Sequence Header                 Assembly
            1       chr|NC_000001|NC_000001.10      hg19
            2       chr|NC_000002|NC_000002.11      hg19
            3       chr|NC_000003|NC_000003.11      hg19
            4       chr|NC_000004|NC_000004.11      hg19
            5       chr|NC_000005|NC_000005.9       hg19
            6       chr|NC_000006|NC_000006.11      hg19
            7       chr|NC_000007|NC_000007.13      hg19
            8       chr|NC_000008|NC_000008.10      hg19
            9       chr|NC_000009|NC_000009.11      hg19
            10      chr|NC_000010|NC_000010.10      hg19
            11      chr|NC_000011|NC_000011.9       hg19
            12      chr|NC_000012|NC_000012.11      hg19
            13      chr|NC_000013|NC_000013.10      hg19
            14      chr|NC_000014|NC_000014.8       hg19
            15      chr|NC_000015|NC_000015.9       hg19
            16      chr|NC_000016|NC_000016.9       hg19
            17      chr|NC_000017|NC_000017.10      hg19
            18      chr|NC_000018|NC_000018.9       hg19
            19      chr|NC_000019|NC_000019.9       hg19
            20      chr|NC_000020|NC_000020.10      hg19
            21      chr|NC_000021|NC_000021.8       hg19
            22      chr|NC_000022|NC_000022.10      hg19
            X       chr|NC_000023|NC_000023.10      hg19
            Y       chr|NC_000024|NC_000024.9       hg19
            M       chr|NC_012920|NC_012920.1       hg19
        </pre>
        
        <h4> 
            <p class="tab">
                Files and paths
            </p>
        </h4>
        
        <p class="tab">
            The tar files <i><font color=blue>genomes_chr_NC_gi.tar.gz</i></font> and <i><font color=blue>genomes_chr_hdr.tar.gz</i></font> include a list of <i><font color=firebrick>chr_NC_gi</i></font> 
            nd <i><font color=firebrick>chr_hdr</i></font> tables for several supported genome assemblies. As mentioned above, this tables are used to convert match lists to BED format.
        </p>
        
        <p class="tab">
            To check the content, type:
        
        <p class="s">
            tar -tvzf genomes_chr_NC_gi.tar.gz<br>
            tar -tvzf genomes_chr_hdr.tar.gz
        </p>
    
        <p class="tab">
            To extract the files, choose a convenient directory (<font color=firebrick>chr_NC_PATH</font>), and execute
            the following commands:
            
        <p class="s">
            cp genomes_chr_NC_gi.tar.gz genomes_chr_hdr.tar.gz chr_NC_PATH/<br>
            cd chr_NC_PATH <br>
            tar -xvzf genomes_chr_NC_gi.tar.gz<br>
            tar -xvzf genomes_chr_hdr.tar.gz
        </p>
        
        <p class="tab">
            The default location (chr_NC_PATH) for conversion programs and scripts is the following:
            <font color=blue><i>/home/local/db/genome</i></font>
            This can be changed via a command option or, more simply, by changing the variable 
            <font color=firebrick>chrNC_dir</font> in the shell wrapper scripts (as described below).
            The CTCF PWM file that will be used in our example is <font color=blue><i>chen10_ctcf.mat</i></font>.
        </p>
        
        
        
        <h3>Analysis Pipeline using two different scanning methods</h3>
        <ul style='list-style-type: none;'>
        <li><b>1) Calculate the matrix score corresponding to p-value=0.00001 (10<sup>-5</sup>)</b></li>
        
        <p class="code">
            matrix_prob -e 0.00001 --bg "0.29,0.21,0.21,0.29" chen10_ctcf.mat
        </p>
            
        <pre>
           SCORE :   1128  PERC :  96.90%
        </pre>
        <p>
           There are two basic methods to scan the genome with a PWM and a cut-off/p-value: </li>
           <ul style='list-style-type: none;'>
               <li>1- Use a fast string-matching algorithm such us <font color=firebrick>bowtie</font> as follows:
               <ul style='list-style-type: circle;'> 
               <li>Given a PWM and a cut-off or p-value, generate all possible matches that represent the PWM with a score greater or equal to the cut-off;
               <li>Map the list of all possible PWM matches to a reference genome ar a set of DNA sequences of interest using a fast string-matching algorithm 
                   (<font color=firebrick>bowtie</font>).
               </ul>
               <li>2- Use <font color=firebrick>matrix_scan</font>, a C program using a conventional search algorithm
            </ul>
           <p> 
           If you are interested in using the first approach, you should read sections 2) and/or 3), whereas for using the conventional algorithm please refer to section 4).
        <p>    
            <li>
                <b>2) Generate a list of all tags that represent all possible PWM matches (with minimal score of 1128)</b><p>
                <font color=firebrick>This step is only necessary if fast string-matchig tools, such as bowtie,
                are used to scan the genome.</font>
            </li>
            
            <p class="code">
                mba -c 1128 -l 19 chen10_ctcf.mat | sort -k2,2 -nr &gt chen10_ctcf_co1128.dat<br><br>
                <font color=firebrick># Convert the tag list into FASTA format (for bowtie):</font><br>
                awk '{print "&gt"$2"\n"$1}' chen10_ctcf_co1128.dat &gt chen10_ctcf_co1128_taglist.fa<br>
            </p>

            The reason for generating a match list in FASTA format is that we want to carry the match score over to the next steps (see below).
           <p>

            <li>
                <b>3) Map the tags representing the PWM to hg19</b>
           <p>
                PWMScan uses the program Bowtie as search engine: 
            </li>
        
           <p> 
                <li>
                    <b><font color=firebrick>1- Bowtie</b></font>
                    <p>
                        Let's define BOWTIE_DIR as the directory containing the bowtie indexed genomes:
                        <ul style='list-style-type: none;'>
                            <li><font color=blue>h_sapiens_hg19</font> identifies the index files linked to the hg19 human assembly</li>
                            <li><font color=blue>BOWTIE_DIR = /home/local/db/bowtie</font></li>
                            <li><font color=blue>chr_NC_PATH = /home/local/db/genome</font></li>
                        </ul>
                    <p>
                        The command to map the list of PWM tags to hg19 is the follwing:
                    <p class="code">
		            <font color=firebrick># It takes of the order of 30 seconds for a full scan of the hg19 genome assembly (with the exclusion of the mitochondrial chromosome). The total number of hits is 143597.</font><br><br> 
                            bowtie -l 19 -n0 -a BOWTIE_DIR/h_sapiens_hg19 -f chen10_ctcf_co1128_taglist.fa --un unmapped.dat &gt chen10_ctcf_co1128_bowtie.out<br>
                            <br>                        
                            bowtie -l 19 -n0 -a /home/local/db/bowtie/h_sapiens_hg19 -f chen10_ctcf_co1128_taglist.fa --un unmapped.dat &gt chen10_ctcf_co1128_bowtie.out<br>
                            <br>
                            <font color=firebrick># Convert the bowtie output to BED via the following command:</font><br>
                            sort -s -k3,3 -k4,4n chen10_ctcf_co1128_bowtie.out | bowtie2bed -s hg19 -l 19 -i chr_NC_PATH &gt chen10_ctcf_co1128_bowtie.bed<br>
                            sort -s -k3,3 -k4,4n chen10_ctcf_co1128_bowtie.out | bowtie2bed -s hg19 -l 19 &gt chen10_ctcf_co1128_bowtie.bed<br>
                            <br>
		            <font color=firebrick># Note: to convert Bowtie output to BED format, we use the chr_hdr file for assembly hg19.<br>
                            # By default, the chr_hdr file is located at chr_NC_PATH/hg19 = /home/local/db/genome/hg19.<br>
                            # chr_NC_PATH can be changed by using option -i.<br>
                            # Also note that the bowtie ouput reports the match score in the first field so that it can be easily retrieved.<br>
                            <br>
                            # To run the bowtie-based pipeline, type:</font><br>
                            bowtie -l 19 -n0 -a BOWTIE_DIR/h_sapiens_hg19 -f chen10_ctcf_co1128_taglist.fa --un unmapped.dat | sort -s -k3,3 -k4,4n | 
                            bowtie2bed -s hg19 -l 19 -i chr_NC_PATH &gt chen10_ctcf_co1128_bowtie.bed<br>
                            <br>
                            <font color=firebrick># Bowtie can read input files from stdin. You should specify "-" for stdin.<br>
                            # In such case, you can run the entire pipeline as follows:</font><br>
                            mba -c 1128 -l 19 chen10_ctcf.mat | sort -k2,2 -nr | awk '{print ">"$2"\n"$1}' | bowtie -l 19 -n0 -a BOWTIE_DIR/h_sapiens_hg19 -f - --un unmapped.dat | sort -s -k3,3 -k4,4n | 
                            bowtie2bed -s hg19 -l 19 -i chr_NC_PATH &gt chen10_ctcf_co1128_bowtie.bed<br>
                            <br>
                            <font color=firebrick># Example:</font><br>
                            bowtie -l 19 -n0 -a /home/local/db/bowtie/h_sapiens_hg19 -f chen10_ctcf_co1128_taglist.fa --un unmapped.dat | sort -s -k3,3 -k4,4n | 
                            bowtie2bed -s hg19 -l 19 &gt chen10_ctcf_co1128_bowtie.bed<br>
                            <br>
                            <font color=firebrick># or</font><br>
                             mba -c 1128 -l 19 chen10_ctcf.mat | sort -k2,2 -nr | awk '{print "&gt"$2"\n"$1}' | bowtie -l 19 -n0 -a /home/local/db/bowtie/h_sapiens_hg19 -f - --un unmapped.dat | 
                            sort -s -k3,3 -k4,4n | bowtie2bed -s hg19 -l 19 &gt chen10_ctcf_co1128_bowtie.bed<br>                           
                        </p>
                    </p>
                </li>
            </p>
        <p>
        <li><b>4) Using a conventional scanning algorithm (matrix_scan)</b> 
        <p>
           # Let's define GENOME_DIR/hg19 as the directory containing all hg19 chromosomes in FASTA files
        <p>
             <ul style='list-style-type: none;'>
             <li><font color=blue>GENOME_DIR = /home/local/db/genome</font>
             <li><font color=blue>chr_NC_PATH = /home/local/db/genome</font>
             </ul>
        <p>    
           # The command to scan the hg19 genome against the chen10 CTCF PWM is the follwing:
        <p>
           <i>cat GENOME_DIR/hg19/chrom*.seq | matrix_scan -m chen10_ctcf.mat -c 1128 | sort -s -k1,1 -k2,2n -k6,6 > chen10_ctcf_co1128_matrix_scan.out</i>
        <p>
           It takes of the order of 3 minutes to scan the entire genome. The total number of hits is 143597 (if we exclude the mitochondrial chromosome) in agreement with the tag mapping methods.<br>
           <font color=firebrick>Note that matrix_scan reports the match score in the fifth field of its output.</font><br>
           <font color=firebrick>Also note that <i>matrix_scan</i> parses the sequence FASTA header assuming that the sequence identifier is of the following type:<br>
        <pre>
        word|Accession|Accession
        Ex:
        chr|NC_000004|NC_000004.11 
        </pre></font>
        <p>
           <p class="code">
           <font color=firebrick># Example:<br></font>
           <i>cat /home/local/db/genome/hg19/chrom*.seq | matrix_scan -m chen10_ctcf.mat -c 1128 | sort -s -k1,1 -k2,2n -k6,6 > chen10_ctcf_co1128_matrix_scan.out</i></font>
        <br>
           <font color=firebrick># Parallelization of matrix_scan<br></font>
           <font color=black>To improve performance, we can run matrix_scan in parallel (on each chromosome file) using a simple perl script called matrix_scan_parallel.pl.
           The above example takes only 30-40 seconds to complete the task, thus competing with the bowtie-based approach.<br>
           To use matrix_scan_parallel.pl, type the following command:<br> </font>
           <i>matrix_scan_parallel.pl /home/local/db/genome/hg19/chrom\*.seq chen10_ctcf.mat 1128 | sort -s -k1,1 -k2,2n -k6,6 &gt chen10_ctcf_co1128_matrix_scan.out</i>
        <br>
           <font color=firebrick># Convert the matrix scan output to BED by issuing the following command:</font>
        <br>
           <i>mscan2bed -s hg19 -i GENOME_DIR chen10_ctcf_co1128_matrix_scan.out > chen10_ctcf_co1128_matrix_scan.bed</i>
        <br>
           <font color=firebrick># Example:<br></font>
           <i>mscan2bed -s hg19 chen10_ctcf_co1128_matrix_scan.out > chen10_ctcf_co1128_matrix_scan.bed</i></font>
        <br>
           <font color=firebrick>
           # Note : to convert matrix_scan output to BED format, we use the chr_NC_gi file for assembly hg19.<br>
           # By default, the chr_NC_gi is located at chr_NC_PATH/hg19 = /home/local/db/genome/hg19.<br> 
             chr_NC_PATH can be changed by using option -i. 
        <br>   
           # To run the entire pipeline:
        <br></font>
           <i>cat GENOME_DIR/hg19/chrom*.seq | matrix_scan -m chen10_ctcf.mat -c 1128 | sort -s -k1,1 -k2,2n -k6,6 | mscan2bed -s hg19 -i GENOME_DIR > chen10_ctcf_co1128_matrix_scan.bed</i>
        <br>
           <font color=firebrick># Example:</font>
        <br>
           <i>cat /home/local/db/genome/hg19/chrom*.seq | matrix_scan -m chen10_ctcf.mat -c 1128 | sort -s -k1,1 -k2,2n -k6,6 | mscan2bed -s hg19 > chen10_ctcf_co1128_matrix_scan.bed</i></font>
        <br>
           <font color=firebrick># Using parallelization:</font>
        <br>
           <i>matrix_scan_parallel.pl /home/local/db/genome/hg19/chrom\*.seq chen10_ctcf.mat 1128 | sort -s -k1,1 -k2,2n -k6,6 | mscan2bed -s hg19 &gt chen10_ctcf_co1128_matrix_scan.bed</i>
        <p>
        </p>
           </li></ul>
           </ul>
        
        </p>

        <p>
        <h3> Shell (bash) Wrappers to execute the analysis pipeline</h3>
         <p class="tab">
         <strong>NOTE: the path for locating the chr_NC_gi/chr_hdr files is hard-coded in the <i>pwm_bowtie_wrapper</i> and <i>pwm_mscan_wrapper</i> scripts (variable chrNC_dir)</strong>
         <p class="tab">
         These shell wrapper scripts have been written to make it easier to run the whole pwmscan pipeline, irrespecitve of what method one chooses.
        <br>
         They require to specify the matrix file containing the integer PWM, the p-value, the path to the assembly or index files, and the UCSC assembly name for the species (e.g. hg19, hg38, mm9 etc.).
        <br>
        <br>
         <strong>A few remarks on the general <i>pwm_scan script</i>.</strong><br>
         The <i>pwm_scan</i> script scans a genome with a PWM using either bowtie or matrix_scan depending on both the p-value and the matrix length.
        <br>
         It basically checks whether the p-value is too high (i.e. the raw score cut-off too low) given the motif lenght to decide on the more suitable search engine.
        <br>
         For high p-values and long motifs, the bowtie-based strategy becomes inefficient because the list of tags representing the PWM and the cut-off becomes too large (of the order of half a billion or more).
        <br>
         The &ltgenome-root-dir&gt argument is the root directory of the genome files (the bowtie index files, and the FASTA chromosome files used by matrix_scan).
        <br>
         &ltgenome-root-dir&gt is supposed to have two sub-directories, bowtie and genome, for storing separately the bowtie indices and the chromosome files respectively.
        <br>
         In our examples, <b>&ltgenome-root-dir&gt</b> is set to <b><i>/home/local/db</i></b>.
        <br>
         The <b>chrNC_dir</b> variable (used for locating the the chr_NC_gi/chr_hdr files) is set to chrNC_dir=<b>&ltgenome_root_dir&gt/genome</b>.
        <br>
         <ul style='list-style-type: none;'>
         <li><b><font color=firebrick>1) Bowtie (pwm_bowtie_wrapper)</b> script</font>
        <p class="code">
           # Usage: `basename $0` &ltmatrix-file&gt &ltp-value&gt &ltbowtie-dir&gt &ltidx-file&gt &ltassembly[hg19|mm9|..]&gt
        <br>
           <font color=firebrick># Example:</font>
        <br>
           <i>pwm_bowtie_wrapper chen10_ctcf.mat 0.00001 /home/local/db/bowtie h_sapiens_hg19 hg19</i></font>
        <br>
           <font color=firebrick># Output file (BED) : chen10_ctcf_co1128_bowtie.bed</font>
        <br>
        </p>

        <li><b><font color=firebrick>2) Matrix Scan (pwm_mscan_wrapper)</b> script</font>
        <p class="code">
           # Usage: `basename $0` &ltmatrix-file&gt &ltp-value&gt &ltgenome-dir&gt &ltassembly[hg19|mm9|..]&gt &ltparallel [0/1]&gt
        <br>
           <font color=firebrick># Example:</font>
        <br>
             <i>pwm_mscan_wrapper chen10_ctcf.mat 0.00001 /home/local/db/genome hg19 1</i></font>
        <br>
           <font color=firebrick># Output file (BED) : chen10_ctcf_co1128_matrix_scan.bed</font>

       <br>
        </p>

        <li><b><font color=firebrick>3) PWM Scan (pwm_scan)</b> script</font>
        <p class="code">
           # Usage: `basename $0` &ltmatrix-file&gt &ltp-value&gt &ltassembly[hg19|mm9|..]&gt &ltgenome-root-dir [ex:/db]&gt &ltparallel [0/1]&gt
        <br>
           <font color=firebrick># Examples:</font>
        <br>
             <i>pwm_scan chen10_ctcf.mat 0.00001 hg19 /home/local/db </i></font>
        <br>
           <font color=firebrick># Or if we set the <parallel> option:</font>
        <br>
             <i>pwm_scan chen10_ctcf.mat 0.00001 hg19 /home/local/db 1</i></font>
        <br>
           <font color=firebrick># Output file (BED) : chen10_ctcf_co1128_matrix_scan.bed</font>
        <br>
           <font color=firebrick># Try also:</font>
        <br>
             <i>pwm_scan chen10_ctcf.mat 0.000001 hg19 /home/local/db 1</i></font>
             <i>pwm_scan chen10_ctcf.mat 0.000001 hg19 /home/local/db </i></font>
        <br>
           <font color=firebrick># Output file (BED) : chen10_ctcf_co1549.bed</font>
        <br>
           <font color=firebrick># N.B.</font>
        <br>
           <font color=black> If the <parallel> option is specified, matrix_scan is run in parallel by splitting the processing of the entire genome
           on multiple processes for each chromosome file in parallel.
        <br>
           Of course it only works on a multi-core processor machine.</font>
        <br>
        </p>
        </ul>


        <h3> How to build Bowtie index files</h3>
        <p>
         <ul style='list-style-type: none;'>
        <p>
          Let's define BOWTIE_DIR as as the directory containing the bowtie indices.
        <p>
          Go to the BOWTIE_DIR directory, and concatenate all chromosome files into a sequence file that includes the entire hg19 genome assembly:
        <p class="code">
          ls -1 -v DB_DIR/chrom*.seq | xargs cat &gt h_sapiens_hg19.fa
        </p>
          You then execute the following command:
        <p class="code">
          bowtie-build -f h_sapiens_hg19.fa h_sapiens_hg19
          </font>
        </p>
    </body>
</html>
