<!DOCTYPE html>
<html lang='en'>
    <head>
        <title>README - Examples</title>
        <style>
            .tab { margin-left: 40px; }
            .code { color: Teal ;
                    font-size: 100%;
                    font-style: italic;
                    line-height: 1.5;
                    padding-left: 20px;
                    border-style: solid;
                    margin-left: 20px;
                    margin-right: 200px;
            }
            .s { color: Teal ;
                    font-size: 100%;
                    font-style: italic;
                    line-height: 1.5;
                    padding-left: 20px;
                    border-style: solid;
                    margin-left: 40px;
                    margin-right: 200px;
            }
        </style>
    </head>

    <body style='background-color:#F2F2F2; font-size:14px; text-align: justify; width:90%'>

        <h1>Application Examples</h1>

        <h2>Scan the hg19 human genome with the chen10 CTCF Position Weight Matrix (PWM)</h2>

        <h3> Important remarks: </h3>
        <p class="tab">
      For the following application examples we will use the hg19 human assembly files that can be found in the <em style="color:blue">genomedb</em> sub-directory of the <a href="https://sourceforge.net/projects/pwmscan" target="_blank"><strong>PWMScan software package</strong></a>.
        <p class="tab">
          The PWMScan programs rely on a few rules to make output conversion to BED format easier, namely FASTA header and file naming conventions.
        <p class="tab">

          The final output format is <strong>BEDdetail</strong>, which is an extension of BED format that is used to enhance the track display page.
          For PWMScan, we use BEDdetail format to include the name of the PWM as well as the p-value associated to the sites identified by PWMScan.

        <p class="tab">
          The <strong>BEDdetail</strong> format provides the following obligatory fields:
        <ul style="list-style: none;">
          <li>1) Chromosome name (e.g. chr1, chrX, etc.)</li>
          <li>2) Starting position of the match</li>
          <li>3) Ending position of the match</li>
          <li>4) Matching nucleotide sequence</li>
          <li>5) Integer score of PWM match</li>
          <li>6) Strand (+, -)</li>
          <li>7) PWM name</li>
          <li>8) P-value of the match</li>
        </ul>
         <p class="tab">

        <h4 class="tab">
            FASTA headers
        </h4>

        <p class="tab">
        All genome sequence files have a FASTA header that is formatted as follows:
        <pre>
            >chr|NC_000001|NC_000001.10 <em>some text</em>
        </pre>

        <p class="tab">
        The <em style="color:blue">sequence identifier</em> includes three words concatenated via a UNIX pipe '|': the word 'chr' followed by the NCBI RefSeq partial and full accession identifiers.
        The accession records indicate sequence identity. The sequence identifier is parsed by most of the PWMScan programs and utilities.
        <p class="tab">
        For each genome assembly, we provide a table for NCBI RefSeq identifier to chromosome number conversion (<em style="color:blue">chr_NC_gi</em>) as well as a table for FASTA sequence identifier to chromosome
        number conversion (<em style="color:blue">chr_hdr</em>). The <em style="color:blue">chr_NC_gi</em> file has been downloaded from NCBI whereas the <em style="color:blue">chr_hdr</em> file has been locally-generated. By consequence, the <em style="color:blue">chr_hdr</em> table can be changed to adapt to other FASTA header format conventions.
        <p class="tab">
        Here is an example for the human genome assembly hg19:
        <pre>
            <em style="color:firebrick">chr_NC_gi</em>
            #Chr    Accession.ver   gi
            1       NC_000001.10    224384768
            2       NC_000002.11    224384767
            3       NC_000003.11    224384766
            4       NC_000004.11    224384765
            5       NC_000005.9     224384764
            6       NC_000006.11    224384763
            7       NC_000007.13    224384762
            8       NC_000008.10    224384761
            9       NC_000009.11    224384760
            10      NC_000010.10    224384759
            11      NC_000011.9     224384758
            12      NC_000012.11    224384757
            13      NC_000013.10    224384756
            14      NC_000014.8     224384755
            15      NC_000015.9     224384754
            16      NC_000016.9     224384753
            17      NC_000017.10    224384752
            18      NC_000018.9     224384751
            19      NC_000019.9     224384750
            20      NC_000020.10    224384749
            21      NC_000021.8     224384748
            22      NC_000022.10    224384747
            X       NC_000023.10    224384746
            Y       NC_000024.9     224384745
            M       NC_012920.1     251831106
        </pre>
        <pre>
            <em style="color:firebrick">chr_hdr</em>
            #Chr    Sequence Header                 Assembly
            1       chr|NC_000001|NC_000001.10      hg19
            2       chr|NC_000002|NC_000002.11      hg19
            3       chr|NC_000003|NC_000003.11      hg19
            4       chr|NC_000004|NC_000004.11      hg19
            5       chr|NC_000005|NC_000005.9       hg19
            6       chr|NC_000006|NC_000006.11      hg19
            7       chr|NC_000007|NC_000007.13      hg19
            8       chr|NC_000008|NC_000008.10      hg19
            9       chr|NC_000009|NC_000009.11      hg19
            10      chr|NC_000010|NC_000010.10      hg19
            11      chr|NC_000011|NC_000011.9       hg19
            12      chr|NC_000012|NC_000012.11      hg19
            13      chr|NC_000013|NC_000013.10      hg19
            14      chr|NC_000014|NC_000014.8       hg19
            15      chr|NC_000015|NC_000015.9       hg19
            16      chr|NC_000016|NC_000016.9       hg19
            17      chr|NC_000017|NC_000017.10      hg19
            18      chr|NC_000018|NC_000018.9       hg19
            19      chr|NC_000019|NC_000019.9       hg19
            20      chr|NC_000020|NC_000020.10      hg19
            21      chr|NC_000021|NC_000021.8       hg19
            22      chr|NC_000022|NC_000022.10      hg19
            X       chr|NC_000023|NC_000023.10      hg19
            Y       chr|NC_000024|NC_000024.9       hg19
            M       chr|NC_012920|NC_012920.1       hg19
        </pre>

        <p class="tab">
        Note that the second field of the <em style="color:blue">chr_hdr</em> file corresponds to the chromosome sequence identifier in the FASTA files. Given that the Bowtie indices are generated starting from the chromosome FASTA files, the Bowtie output will include the sequence identifier as specified in the second field of the <em style="color:blue">chr_hdr</em> file. To convert to BED format, which uses chromosome numbers, programs such as <em style="color:blue">bowtie2bed</em> need to read the <em style="color:blue">chr_hdr </em> table.
        <p class="tab">
        If Bowtie indices are generated from chromosome FASTA files with different headers (or sequence identifiers), the second field of the <em style="color:blue">chr_hdr</em> table has to be changed accordingly.

        <p class="tab">
        Chromosome files downloaded from UCSC are named chr*.fa. If this is the case, the adapted wrapper scripts <em>pwm_scan_ucsc</em> and <em>pwm_mscan_wrapper_ucsc</em> should be used instead of the deafult ones <em>pwm_scan</em> and <em>pwm_mscan_wrapper</em> (see the <strong>Wrapper Scripts</strong> Section). The <em>pwm_bowtie_wrapper</em> script should in principle work with genome index files built from UCSC genome assemblies, provided the <em style="color:blue">chr_hdr</em> file is changed according to the UCSC chromosome header convention, which is the following:
        <pre>
            >chr1
        </pre>
        <p class="tab">
        An example of such header files for the mouse genome mm10 downloaded from UCSC would then be the following:

        <pre>
            <em style="color:firebrick">chr_NC_gi</em>
             #Chr    Accession.ver   gi      Assembly
             1       chr1    372099109       mm10
             2       chr2    372099108       mm10
             3       chr3    372099107       mm10
             4       chr4    372099106       mm10
             5       chr5    372099105       mm10
             6       chr6    372099104       mm10
             7       chr7    372099103       mm10
             8       chr8    372099102       mm10
             9       chr9    372099101       mm10
             10      chr10   372099100       mm10
             11      chr11   372099099       mm10
             12      chr12   372099098       mm10
             13      chr13   372099097       mm10
             14      chr14   372099096       mm10
             15      chr15   372099095       mm10
             16      chr16   372099094       mm10
             17      chr17   372099093       mm10
             18      chr18   372099092       mm10
             19      chr19   372099091       mm10
             X       chrX    372099090       mm10
             Y       chrY    372099089       mm10
             M       chrM    34538597        mm10
        </pre>
        <pre>
            <em style="color:firebrick">chr_hdr</em>
             #Chr    Sequence Header  Assembly
             1       chr1             mm10
             2       chr2             mm10
             3       chr3             mm10
             4       chr4             mm10
             5       chr5             mm10
             6       chr6             mm10
             7       chr7             mm10
             8       chr8             mm10
             9       chr9             mm10
             10      chr10            mm10
             11      chr11            mm10
             12      chr12            mm10
             13      chr13            mm10
             14      chr14            mm10
             15      chr15            mm10
             16      chr16            mm10
             17      chr17            mm10
             18      chr18            mm10
             19      chr19            mm10
             X       chrX             mm10
             Y       chrY             mm10
             M       chrM             mm10
        </pre>

        <h4 class="tab">
                Files and paths
        </h4>

        <p class="tab">
            The tar files <em style="color:blue">genomes_chr_NC_gi.tar.gz</em> and <em style="color:blue">genomes_chr_hdr.tar.gz</em> include a list of <em style="color:firebrick">chr_NC_gi</em> and <em style="color:firebrick">chr_hdr</em> tables for several common model organisms such as human, mouse, fruit fly, worm, zebrafish, yeast, and more. As mentioned above, this tables are used to convert match lists to BED format.
        </p>

        <p class="tab">
            To check the content, type:

        <p class="s">
            tar -tvzf genomes_chr_NC_gi.tar.gz<br>
            tar -tvzf genomes_chr_hdr.tar.gz
        </p>

        <p class="tab">
            To extract the files, execute the following commands:

        <p class="s">
            cd genomedb<br>
            tar -xvzf genomes_chr_NC_gi.tar.gz<br>
            tar -xvzf genomes_chr_hdr.tar.gz
        </p>

        <p class="tab">
            The default genome root location (variable genome_root_dir) for conversion programs and scripts is the following:<br>
        <p class="tab">
            <span style="color:blue">genome_root_dir=<em>../genomedb</em></span><br>
        <p class="tab">
            which corresponds to the root directory of the entire genome assembly data, and is the genome directory used for the following application examples..<br>
        <p class="tab">
            The genome root directory can be changed via the command option -d <span style="color:firebrick">&lt;genome-root-dir&gt;</span> in all bash wrapper scripts (as described below).<br>
        <p class="tab">
            The PWM for the CTCF transcription factor that will be used in our example is the <em style="color:blue">chen10_ctcf.mat</em> file in the <em style="color:blue">examples</em> sub-directory..
        <p class="tab">
Upon installation, all PWMScan-specific binary files are located in the bin sub-directory, so we define:<br>
        <p class="tab">
            <span style="color:blue">bin_dir=<em>../bin</em></span>
        </p>



        <h3>Analysis Pipeline using two different scanning methods</h3>
        <h4><strong>1) Calculate the matrix score corresponding to p-value=0.00001 (10<sup>-5</sup>)</strong></h4>

        <p class="code">
            ../bin/matrix_prob -e 0.00001 --bg "0.29,0.21,0.21,0.29" chen10_ctcf.mat
        </p>

        <pre>
           SCORE :   1128  PERC :  96.90%
        </pre>
        <p>
           There are two basic methods to scan the genome with a PWM and a cut-off/p-value:
           <ul style='list-style-type: none;'>
               <li>1- Use a fast string-matching algorithm such us <span style="color:firebrick">Bowtie</span> as follows:
               <ul style='list-style-type: circle;'>
               <li>Given a PWM and a cut-off or p-value, generate all possible matches that represent the PWM with a score greater or equal to the cut-off;
               <li>Map the list of all possible PWM matches to a reference genome ar a set of DNA sequences of interest using Bowtie.
                   (<span style="color:firebrick">Bowtie</span>).
               </ul>
               <li>2- Use <span style="color:firebrick">matrix_scan</span>, a C program using a conventional search algorithm that has been optimized for rapid score computation and drop-off strategy.
            </ul>
           <p>
           If you are interested in using the first approach, you should read <strong>Sections 2), 3) and 4)</strong>, whereas for using matrix_scan please refer to <strong>Sections 2) and 5)</strong>.

            <h4><strong>2) Generate the PWM Score Cumulative Distribution</strong></h4>
                The cumulative score distribution is used to assign a p-value to a match score.
                The matrix_prob program is used to compute the cumulative score distribution.

            <p class="code">
                ../bin/matrix_prob --bg "0.29,0.21,0.21,0.29" chen10_ctcf.mat &gt; chen10_ctcf_co1128_scoretab.txt
            </p>

                <h4><strong>3) Generate a list of all tags that represent all possible PWM matches (with minimal score of 1128)</strong></h4>
                <span style="color:firebrick">This step is only necessary if fast string-matchig tools, such as Bowtie,
                are used to scan the genome.</span>

            <p class="code">
                ../bin/mba -c 1128 chen10_ctcf.mat &gt; chen10_ctcf_co1128.dat<br>
                <span style="color:firebrick"># Convert the tag list into FASTA format (for Bowtie):</span><br>
                awk '{print "&gt;"$2"\n"$1}' chen10_ctcf_co1128.dat &gt; chen10_ctcf_co1128_taglist.fa<br>
            </p>

            The reason for generating a match list in FASTA format is that we want to carry the match score over to the next steps (see below).

                <h4><strong>4) Map the tags representing the PWM to hg19</strong></h4>
                PWMScan uses the program Bowtie to map all matching tags to the hg19 genome assembly.

           <p>
                    <strong style="color:firebrick">1- Bowtie</strong>
                    <p>
                       Let's define GENOME_DIR as the genome root directory:
                        <ul style='list-style-type: none;'>
                            <li style="color:blue">GENOME_DIR = ../genomedb</li>
                            <li style="color:blue">chr_NC_PATH = GENOME_DIR</li>
                        </ul>
                    <p>
                        Let's define BOWTIE_DIR as the directory containing the Bowtie indexed genomes:
                        The <em style="color:blue">h_sapiens_hg19</em> tag identifies the Bowtie index files linked to the hg19 human assembly.<br>
                        To build the h_sapiens_hg19 index file for the hg19 assembly, please follow the instructions given in the last Section of this README page (<strong>How to build Bowtie index files</strong>).
                        <ul style='list-style-type:none'>
                            <li style="color:blue">BOWTIE_DIR = GENOME_DIR/bowtie</li>
                        </ul>
                    <p>
                        The command to map the list of PWM tags to hg19 is the following:
                    <p>
                        <em>bowtie -l 19 -n0 -a BOWTIE_DIR/h_sapiens_hg19 -f chen10_ctcf_co1128_taglist.fa --un unmapped.dat &gt; chen10_ctcf_co1128_bowtie.out</em>
                    <p>
                        It takes of the order of 12-15 seconds for a full scan of the hg19 genome assembly (with the exclusion of the mitochondrial chromosome). The total number of hits is 143597.<br>

                    <p class="code">
                            <span style="color:firebrick"># Example:</span>
                            <br>
                            bowtie --threads 4 -l 19 -n0 -a ../genomedb/bowtie/h_sapiens_hg19 -f chen10_ctcf_co1128_taglist.fa --un unmapped.dat &gt;chen10_ctcf_co1128_bowtie.out<br>
                            <span style="color:firebrick"># Convert the bowtie output to <strong style='font-size:13px'>BEDdetail</strong> via the following command:</span><br>
                            sort -s -k3,3 -k4,4n chen10_ctcf_co1128_bowtie.out | ../bin/bowtie2bed -s hg19 -l 19 -i chr_NC_PATH | awk 'BEGIN { while((getline line &lt; "chen10_ctcf_co1128_scoretab.txt") &gt; 0 ) {split(line,f," "); pvalue[f[1]]=f[2]} close("chen10_ctcf_co1128_scoretab.txt")} {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t""chen10_ctcf""\t""P-value="pvalue[$5]}' &gt; chen10_ctcf_co1128_bowtie.bed<br>

                            <span style="color:firebrick"># Example (use default chr_NC_PATH):</span><br>
                            sort -s -k3,3 -k4,4n chen10_ctcf_co1128_bowtie.out | ../bin/bowtie2bed -s hg19 -l 19 -i ../genomedb | awk 'BEGIN { while((getline line &lt; "chen10_ctcf_co1128_scoretab.txt") &gt; 0 ) {split(line,f," "); pvalue[f[1]]=f[2]} close("chen10_ctcf_co1128_scoretab.txt")} {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t""chen10_ctcf""\t""P-value="pvalue[$5]}' &gt; chen10_ctcf_co1128_bowtie.bed<br>
                            <br>
                            <span style="color:firebrick"># Note: to convert Bowtie output to BED format, we use the chr_hdr file for assembly hg19.<br>
                            # By default, the chr_hdr file is located at chr_NC_PATH/hg19 = ../genomedb/hg19.<br>
                            # chr_NC_PATH can be changed by using option -i.<br>
                            # Also note that the bowtie ouput reports the match score in the first field so that it can be easily retrieved.<br>
                            <br>
                            # Run the Bowtie-based pipeline:<br>
                            # Bowtie can read input files from stdin. You should specify "-" for stdin.<br>
                            # In such case, you can run the entire pipeline as follows:</span><br>
                             ../bin/mba -c 1128 chen10_ctcf.mat | awk '{print "&gt;"$2"\n"$1}' | bowtie --threads 4 -l 19 -n0 -a ../genomedb/bowtie/h_sapiens_hg19 -f - --un unmapped.dat | sort -s -k3,3 -k4,4n | ../bin/bowtie2bed -s hg19 -l 19 -i ../genomedb | awk 'BEGIN { while((getline line &lt; "chen10_ctcf_co1128_scoretab.txt") &gt; 0 ) {split(line,f," "); pvalue[f[1]]=f[2]} close("chen10_ctcf_co1128_scoretab.txt")} {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t""chen10_ctcf""\t""P-value="pvalue[$5]}' &gt; chen10_ctcf_co1128_bowtie.bed<br>
                        </p>

        <h4><strong>5) Using matrix_scan</strong></h4>
        <p>
           Let's define GENOME_DIR as the genome root directory:
           <ul style='list-style-type:none'>
             <li style="color:blue">GENOME_DIR = ../genomedb</li>
             <li style="color:blue">chr_NC_PATH = GENOME_DIR</li>
           </ul>

           We define as ASSEMBLY_DIR the sub-directory containing all the hg19 chromosome FASTA files:
           <ul style='list-style-type:none'>
             <li style="color:blue">ASSEMBLY_DIR = GENOME_DIR/hg19</li>
           </ul>
        <p>
           The command to scan the hg19 genome against the chen10 CTCF PWM is the following:
        <p>
           <em>cat ASSEMBLY_DIR/chrom*.seq | ../bin/matrix_scan -m chen10_ctcf.mat -c 1128 | sort -s -k1,1 -k2,2n -k6,6 > chen10_ctcf_co1128_matrix_scan.out</em>
        <p>
           It takes of the order of 30 seconds to scan the entire genome. The total number of hits is 143597 (if we exclude the mitochondrial chromosome) in agreement with the tag mapping methods.<br>
           <span style="color:firebrick">Note that matrix_scan reports the match score in the fifth field of its output.</span><br>
           <span style="color:firebrick">Also note that <em>matrix_scan</em> parses the sequence FASTA header assuming that the sequence identifier is of the following type:</span><br>
        <pre>
        word|Accession|Accession
        Ex:
        chr|NC_000004|NC_000004.11
        </pre>
        <p>
           <span style="color:firebrick"># Parallelization of matrix_scan</span><br>
           To improve performance, we can run matrix_scan in parallel (on each chromosome file) using <strong>GNU parallel</strong> that dispatches the jobs to available CPU-cores. In this way, matrix_scan competes with the Bowtie-based approach.
         <p class="code">
           <span style="color:firebrick"># Example:</span><br>
           <em>cat ../genomedb/hg19/chrom*.seq | ../bin/matrix_scan -m chen10_ctcf.mat -c 1128 | sort -s -k1,1 -k2,2n -k6,6 &gt; chen10_ctcf_co1128_matrix_scan.out</em>
        <br>
           <span style="color:firebrick"># To use GNU parallel, type the following command:</span><br>
           <em>find ../genomedb/hg19/ -name chrom\*.seq -exec echo "../bin/matrix_scan -m chen10_ctcf.mat -c 1128" {} \; | grep -v chromMt | parallel -P 15 | sort -s -k1,1 -k2,2n -k6,6 &gt; chen10_ctcf_co1128_matrix_scan.out</em>
        <br><br>
           <span style="color:firebrick"># The -P option is used to set the maximum number of parallel processes for execution of matrix_scan. The above example takes only about 20 seconds to complete the task in agreement with Bowtie.</span><br><br>
           # Convert the matrix scan output to <strong style='font-size:13px'>BEDdetail</strong> by issuing the following command:
        <br>
           <span style="color:firebrick"># Example:</span><br>
           <em>../bin/mscan2bed -s hg19 -i ../genomedb chen10_ctcf_co1128_matrix_scan.out | awk 'BEGIN { while((getline line &lt; "chen10_ctcf_co1128_scoretab.txt") &gt; 0 ) {split(line,f," "); pvalue[f[1]]=f[2]} close("chen10_ctcf_co1128_scoretab.txt")} {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t""chen10_ctcf""\t""P-value="pvalue[$5]}' &gt; chen10_ctcf_co1128_matrix_scan.bed</em>
        <br><br>
           <span style="color:firebrick">
           # Note : to convert matrix_scan output to BED format, we use the chr_NC_gi file for assembly hg19.<br>
           # By default, the chr_NC_gi is located at chr_NC_PATH/hg19 = ../genomedb/hg19.<br>
             &nbsp;&nbsp;chr_NC_PATH can be changed by using the &lt;-i&gt; option.
        <br><br>
           # To run the entire pipeline:</span>
        <br>
           <em>cat ../genomedb/hg19/chrom*.seq | ../bin/matrix_scan -m chen10_ctcf.mat -c 1128 | sort -s -k1,1 -k2,2n -k6,6 | ../bin/mscan2bed -s hg19 -i ../genomedb | awk 'BEGIN { while((getline line &lt; "chen10_ctcf_co1128_scoretab.txt") &gt; 0 ) {split(line,f," "); pvalue[f[1]]=f[2]} close("chen10_ctcf_co1128_scoretab.txt")} {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t""chen10_ctcf""\t""P-value="pvalue[$5]}' &gt; chen10_ctcf_co1128_matrix_scan.bed</em>
        <br>
           <span style="color:firebrick"># Using parallelization:</span>
        <br>
           <em>find ../genomedb/hg19/ -name chrom\*.seq -exec echo "../bin/matrix_scan -m chen10_ctcf.mat -c 1128" {} \; | grep -v chromMt | parallel -P 15 | sort -s -k1,1 -k2,2n -k6,6 | ../bin/mscan2bed -s hg19 -i ../genomedb | awk 'BEGIN { while((getline line &lt; "chen10_ctcf_co1128_scoretab.txt") &gt; 0 ) {split(line,f," "); pvalue[f[1]]=f[2]} close("chen10_ctcf_co1128_scoretab.txt")} {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t""chen10_ctcf""\t""P-value="pvalue[$5]}' &gt; chen10_ctcf_co1128_matrix_scan.bed</em>
        <p>
        <br>
        </p>
           </li></ul>

        <h3> Shell (bash) Wrappers to execute the analysis pipeline</h3>
         <p class="tab">
         These shell wrapper scripts have been written to make it easier to run the whole pwmscan pipeline, irrespecitve of what method one chooses.
        <br>
         They require to specify the matrix file containing the integer PWM, the p-value, the path to the assembly or index files, and the UCSC assembly name for the species (e.g. hg19, hg38, mm9 etc.).
         Optional parameters are the overlapping matches flag (-o), the forward scanning (-f) option, the background base composition (-b), the parallel execution (-p) and write to file (-w) flags.
        <br>
        If the '-o' option is specified, overlapping matches are retained. The '-f' option activates forward scanning. If the '-p' option is set, the matrix_scan program execution is done by parallel dispatch. If the -w option is set, the results are written to file, else output goes to STDOUT.
        <br>
        The output file is a match list in <strong>BEDdetail</strong> format.
        <br>
        <br>
        The <strong>pwm_scan</strong> script scans a genome with a PWM using either Bowtie or matrix_scan depending on both the p-value and the matrix length.
        For high p-values and long motifs, the Bowtie-based strategy becomes inefficient and the matrix_scan-based approach is the fastest one.
        <br>
        The <strong>pwm_bowtie_wrapper</strong> script implements the Bowtie-based pipeline whereas the <strong>pwm_mscan_wrapper</strong> script uses the matrix_scan-based approach.
        <br>
        The <strong>pwm_scan_ucsc</strong> and <strong>pwm_mscan_wrapper_ucsc</strong> scripts are just adaptations of the original ones in order to deal with chromosome files downloaded from UCSC.
        <br>
        <br>
        The <strong>pwmlib_scan</strong> script scans a genome with a collection of PWMs coming from a database.<br>
        Accepted motif database formats are the MEME libraries as well as the motif libraries (both log-odds and letter-probability formats) provided by the PWMScan Web interface.

         <p class="tab">
         <strong>NOTE:</strong> The path to all the PWMScan binaries and scripts is defined by the <strong>bin_dir</strong> variable and is hard-coded in all bash scripts, and is changed upon installation via the Makefile, so that all installed scripts in <strong>bin_dir</strong> have the correct binary pathname.

        <p class="tab" style='font-weight: bold; font-size:13px'>
         A few remarks on the general <em>pwm_scan script</em>.
        <p class="tab">
         The &lt;genome-root-dir&gt; input argument is the root directory of the genome files (the Bowtie index files, and the FASTA chromosome files used by matrix_scan, and the assembly tables used to convert the PWMScan pipeline output to BED format).
        <br>
         &lt;genome-root-dir&gt; is supposed to have two sub-directories types, the /bowtie sub-directory where the Bowtie index files are stored, and the /&lt;assembly&gt; sub-directory for storing assembly-specific chromosome files.
        <p class="tab">
         In our setting, <strong>&lt;genome-root-dir&gt;</strong> is set to <span style='font-weight: bold; font-size:13px'><em>../genomedb</em></span>.
        <br>
         The <strong>chrNC_dir</strong> variable (used for locating the the chr_NC_gi/chr_hdr files) is set to chrNC_dir=<strong>&lt;genome-root-dir&gt;</strong>.
        <br>
         <ul style='list-style-type: none;'>
         <li><strong style="color:firebrick">1) Bowtie (pwm_bowtie_wrapper)</strong> script
        <p class="code">
           # Usage: <br>pwm_bowtie_wrapper -m &lt;matrix-file&gt; -e &lt;p-value&gt; -d &lt;genome-root-dir&gt; -s &lt;assembly[hg19|mm9|..]&gt; [options]<br>
        - version 1.1.6<br>
        where options are:<br>
             -b &lt;bg_comp&gt; &nbsp;&nbsp; Background model (residue priors), e.g. : 0.29,0.21,0.21,0.29<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[def=predefined for some common genome assemblies, else set to uniform]<br>
             -f  &nbsp;&nbsp;&nbsp; Scan sequences in forward direction [def=bidirectional]<br>
             -o  &nbsp;&nbsp;&nbsp; Allow for overlapping matches [def=non-overlapping matches]<br>
             -w  &nbsp;&nbsp;&nbsp; Write output to file [def=STDOUT]<br>
        <br>
           <span style="color:firebrick"># Example:</span>
        <br>
           <em>../bin/pwm_bowtie_wrapper -m chen10_ctcf.mat -e 0.00001 -d ../genomedb -s hg19 -o -w</em>
        <br>
           <span style="color:firebrick"># Or:</span>
        <br>
           <em>bash ../bin/pwm_bowtie_wrapper -m chen10_ctcf.mat -e 0.00001 -d ../genomedb -s hg19 -o -w</em>
        <br>
           <span style="color:firebrick"># List of matches (BED format) : chen10_ctcf_co1128_bowtie.bed</span>
        <br>
           <span style="color:firebrick"># Total number of PWM matches  : 143597</span>
        <br>
        </p>

        </li><li><strong style="color:firebrick">2) Matrix Scan (pwm_mscan_wrapper)</strong> script
        <p class="code">
           # Usage:<br> pwm_mscan_wrapper -m &lt;matrix-file&gt; -e &lt;p-value&gt; -d &lt;genome-root-dir&gt; -s &lt;assembly[hg19|mm9|..]&gt; [options]<br>
        - version 1.1.6<br>
        where options are:<br>
             -b &lt;bg_comp&gt; &nbsp;&nbsp; Background model (residue priors), e.g. : 0.29,0.21,0.21,0.29<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[def=predefined for some common genome assemblies, else set to uniform]<br>
             -f  &nbsp;&nbsp;&nbsp; Scan sequences in forward direction [def=bidirectional]<br>
             -o  &nbsp;&nbsp;&nbsp; Allow for overlapping matches [def=non-overlapping matches]<br>
             -w  &nbsp;&nbsp;&nbsp; Write output to file [def=STDOUT]<br>
             -p  &nbsp;&nbsp;&nbsp; Distribute matrix_scan on multiple CPU-cores [def=non-parallel processing]<br>
        <br>
           <span style="color:firebrick"># Example:</span>
        <br>
             <em>../bin/pwm_mscan_wrapper -m chen10_ctcf.mat -e 0.00001 -d ../genomedb -s hg19 -o -p -w</em>
        <br>
           <span style="color:firebrick"># Or:</span>
        <br>
             <em>bash ../bin/pwm_mscan_wrapper -m chen10_ctcf.mat -e 0.00001 -d ../genomedb -s hg19 -o -p -w</em>
        <br>
           <span style="color:firebrick"># List of matches (BED format) : chen10_ctcf_co1128_matrix_scan.bed</span>
        <br>
           <span style="color:firebrick"># Total number of PWM matches  : 143597</span>
       <br>
        </p>

        </li><li><strong style="color:firebrick">3) PWM Scan (pwm_scan)</strong> script
        <p class="code">
           # Usage:<br> pwm_scan -m &lt;matrix-file&gt; -e &lt;p-value&gt; -d &lt;genome-root-dir&gt; -s &lt;assembly[hg19|mm9|..]&gt; [options]<br>
        - version 1.1.6<br>
        where options are:<br>
             -b &lt;bg_comp&gt; &nbsp;&nbsp; Background model (residue priors), e.g. : 0.29,0.21,0.21,0.29<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[def=predefined for some common genome assemblies, else set to uniform]<br>
             -f  &nbsp;&nbsp;&nbsp; Scan sequences in forward direction [def=bidirectional]<br>
             -o  &nbsp;&nbsp;&nbsp; Allow for overlapping matches [def=non-overlapping matches]<br>
             -w  &nbsp;&nbsp;&nbsp; Write output to file [def=STDOUT]<br>
             -p  &nbsp;&nbsp;&nbsp; Distribute matrix_scan on multiple CPU-cores [def=non-parallel processing]<br>
        <br>
           <span style="color:firebrick"># Examples:</span>
        <br>
             <em>bash ../bin/pwm_scan -m chen10_ctcf.mat -e 0.00001 -d ../genomedb -s hg19 -o -w</em>
        <br>
           <span style="color:firebrick"># List of matches (BED format) : chen10_ctcf_co1128_bowtie.bed</span>
        <br>
           <span style="color:firebrick"># Or if we set the &lt;parallel&gt; option:</span>
        <br>
             <em>bash ../bin/pwm_scan -m chen10_ctcf.mat -e 0.00001 -d ../genomedb -s hg19 -o -w -p</em>
        <br>
           <span style="color:firebrick"># List of matches (BED format) : chen10_ctcf_co1128_matrix_scan.bed</span>
        <br>
           <span style="color:firebrick"># Also try the following commands (with p-value=10<sup>-6</sup>):</span>
        <br>
             <em>bash ../bin/pwm_scan -m chen10_ctcf.mat -e 0.000001 -d ../genomedb -s hg19 -w -p</em>
        <br>
           <span style="color:firebrick"># List of matches (BED format) : chen10_ctcf_co1549_bowtie.bed</span>
        <br>
             <em>bash ../bin/pwm_scan -m stat1.mat -e 0.000001 -d ../genomedb -s hg19 -o -w -p</em>
        <br>
           <span style="color:firebrick"># List of matches (BED format) : stat1_co1731_bowtie.bed</span>
        <br>
           <span style="color:firebrick"># N.B.</span>
        <br>
           <span style="color:black">If the &lt;parallel&gt; option is set (default), matrix_scan is run in parallel by splitting the processing of the entire genome on multiple processes for each chromosome file in parallel.
        <br>
           Of course, this only works on a multi-core processor machine.</span>
        <br>
        </p>

        </li><li><strong style="color:firebrick">4) PWMlib Scan (pwmlib_scan)</strong> script
        <p class="code">
           # Usage:<br> pwmlib_scan -l &lt;pwmlib-file&gt; -e &lt;p-value&gt; -d &lt;genome-root-dir&gt; -s &lt;assembly[hg19|mm9|..]&gt; [options]<br>
        - version 1.1.6<br>
        where options are:<br>
             -b &lt;bg_comp&gt; &nbsp;&nbsp; Background model (residue priors), e.g. : 0.29,0.21,0.21,0.29<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[def=predefined for some common genome assemblies, else set to uniform]<br>
             -g  &nbsp;&nbsp;&nbsp; PWM format. Accepted formats are : meme, logodds [def], and lpm<br>
             -f  &nbsp;&nbsp;&nbsp; Scan sequences in forward direction [def=bidirectional]<br>
             -o  &nbsp;&nbsp;&nbsp; Allow for overlapping matches [def=non-overlapping matches]<br>
             -w  &nbsp;&nbsp;&nbsp; Write output to file [def=STDOUT]<br>
             -p  &nbsp;&nbsp;&nbsp; Distribute matrix_scan on multiple CPU-cores [def=non-parallel processing]<br>
             -f  &nbsp;&nbsp;Scan sequences in forward direction [def=bidirectional]<br>
             -o  &nbsp;&nbsp;Allow for overlapping matches [def=non-overlapping matches]<br>
             -w  &nbsp;&nbsp;Write output to file [def=STDOUT]<br>
             -p  &nbsp;&nbsp;Distribute matrix_scan on multiple CPU-cores [def=non-parallel processing]<br>
             -v  &nbsp;&nbsp;Verbose mode and keep PWM tmp files<br>
        <br>
           <span style="color:firebrick"># Example:</span>
        <br>
           <span style="color:firebrick"># Scan the human genome hg19 with the Isakova 2017 collection (lpm format):</span>
        <br>
             <em>bash ../bin/pwmlib_scan -l ../pwmlibs/isakova2017_human_matrix_probs.mat -g lpm -e 0.00001 -d ../genomedb -s hg19 -p 2&gt;pwmlib_scan_log.txt | sort -s -k1,1 -k2,2n &gt; isakova2017_human_hg19_pwmscan.bed</em>
        <br>
           <span style="color:firebrick"># On a linux multi-core platform with 48 CPU cores, the whole process takes about 10 minutes for scanning the entire human genome with a total of 86 matrices.</span>
        <br>
        <br>
          </li></ul>

        <h3> Shell (bash) Wrapper for matrix conversion</h3>
         <p class="tab">
         The pwm_convert bash script is used to convert JASPAR, TRANSFAC, PFM, LPM and real PWM formats to integer log-odds format.
        <div class="code">
           # Usage: pwm_convert &lt;matrix-file&gt; -f=&lt;format [jaspar|transfac|lpm|pfm|real]&gt;
             <br>
 &nbsp;&nbsp; OPTIONS:
             <br>
&nbsp;&nbsp;&nbsp;&nbsp;  -h                  show help message
             <br>
&nbsp;&nbsp;&nbsp;&nbsp;  -b=&lt;bg-freq-file&gt;   bg nucleotide composition file
             <br>
&nbsp;&nbsp;&nbsp;&nbsp;  -c=&lt;pseudo-cnt&gt;     pseudo-count fraction [def=0]
             <br>
&nbsp;&nbsp;&nbsp;&nbsp;  -m=&lt;minscore&gt;       minimal score value [def=-10000]
             <br>
&nbsp;&nbsp;&nbsp;&nbsp;  -n=&lt;log-scaling&gt;    log scaling factor [def=100]
             <br>
&nbsp;&nbsp;&nbsp;&nbsp;  -s=&lt;mult&gt;           matrix multiplication factor (for real PWMs) [def=100]
             <br>
&nbsp;&nbsp;&nbsp;&nbsp;  Convert JASPAR, TRANSFAC, PFM, LPM and real PWM formats to integer log likelihoods (log-odds) format.
        <br>
        <span style="color:black">By default, the prior background nucleotide frequencies are set to 0.25.</span><br>
        <br>
        <span style="color:firebrick"># Examples:</span>
        <br>
        <span style="color:firebrick"># Convert JASPAR to log-odds:</span>
        <br>
             <em>bash ../bin/pwm_convert stat1_jaspar.mat -f=jaspar 2&gt;/dev/null</em><br>

        <span style="color:firebrick"># If we want to use different backgroung nucleotide frequencies:</span>
        <br>
             <em>bash ../bin/pwm_convert stat1_jaspar.mat -f=jaspar -b=bg_probs_2.txt 2&gt;/dev/null</em><br>

        <span style="color:black"># Where the bg_probs_2.txt file includes the prior bg frequencies, defined as follows:
        <br>
         &nbsp;&nbsp;A  0.29 <br>
         &nbsp;&nbsp;C  0.21 <br>
         &nbsp;&nbsp;G  0.21 <br>
         &nbsp;&nbsp;T  0.29 <br></span>

        <span style="color:firebrick"># Convert TRANSFAC to log-odds:</span>
        <br>
             <em>bash ../bin/pwm_convert statx_transfac.mat -f=transfac -c=0.0000001 2&gt;/dev/null</em><br>

        <span style="color:firebrick"># Convert PFM (Position Frequency Matrix) to log-odds:</span>
        <br>
             <em>bash ../bin/pwm_convert stat1_pfm.mat -f=pfm 2&gt;/dev/null</em><br>
        <br>
        </div>

        <h3> How to compute the background base composition of a set of DNA sequences</h3>
        <p class="tab">
        The program <em style="color:blue">seq_extract_bcomp</em> can be used to extract BED regions from a set of FASTA-formatted sequences or, optionally, compute the base composition of a set of DNA sequences. When computing the base composition, the seq_extract_bcomp program can either extract BED-defined regions or use the FASTA input as a whole (if the BED file is not given).<br><br>

        <em># Usage: seq_extract_bcomp [options] [-f &lt;bed_file&gt;] [-s &lt;species&gt;] [&lt;] [&lt;fasta_file&gt;|stdin]<br><br>
     where options are:<br>
&nbsp;&nbsp; -d          &nbsp;&nbsp;&nbsp;&nbsp;Print debug information<br>
&nbsp;&nbsp; -h          &nbsp;&nbsp;&nbsp;&nbsp;Show this help text<br>
&nbsp;&nbsp; -c          &nbsp;&nbsp;&nbsp;&nbsp;Compute base composition [def=on forward strand]<br>
&nbsp;&nbsp; -b          &nbsp;&nbsp;&nbsp;&nbsp;Compute base composition on both strand [-c is required]<br>
&nbsp;&nbsp; -r          &nbsp;&nbsp;&nbsp;&nbsp;Compute base composition on reverse strand [-c is required]<br>
&nbsp;&nbsp; -i &lt;int&gt;    &nbsp;&nbsp;&nbsp;&nbsp;AC index (after how many pipes |) for FASTA header [2]<br>
&nbsp;&nbsp; -p &lt;path&gt;   &nbsp;&nbsp;&nbsp;&nbsp;Use &lt;path&gt; to locate the chr_NC_gi file [if BED file is given]<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[default is: $HOME/db/genome]<br><br>

        Extract BED regions from a set of FASTA-formatted sequences.<br>
        The extracted sequences are written to standard output.<br>
        Optionally (-c), the program computes and only outputs the base composition,<br>
        in which case sequences can be extracted directly from the FASTA input or,<br>
        as for the extraction mode, specified in a BED file. If the BED file is not<br>
        given, the &lt;species&gt; argument is not required. If base composition mode is set<br>
        (-c option), the program can optionally compute it on both strands (-b option)<br>
        for strand-symmetric base composition or on the reverse strand only (-r).</em><br><br>
        <p class="code">
        <span style="color:firebrick"># Here are a few examples of use.</span><br>
        <span style="color:firebrick"># Extract STAT1 peak regions (16158 peaks) defined in the BED file Encode_Helas3Stat1_peaks.bed from the human assembly hg19</span>
        <br>
           <em>cat ../genomedb/hg19/chrom[^Mt]*.seq | ../bin/seq_extract_bcomp -p ../genomedb -f Encode_Helas3Stat1_peaks.bed -s hg19 > seq_extract_stat1_peaks.fa</em><br>
        <span style="color:firebrick"># This command takes about 20 seconds.</span><br>
        <span style="color:firebrick"># Compute the base composition of the STAT1 peak regions on the forward strand</span>
        <br>
           <em> ../bin/seq_extract_bcomp -c seq_extract_stat1_peaks.fa</em><br><br>
            Total Sequence length: 4863558<br>
            0.2445,0.2547,0.2544,0.2464<br>
        <span style="color:firebrick"># Compute the base composition of the STAT1 peak regions on both strands</span>
        <br>
           <em> ../bin/seq_extract_bcomp -cb seq_extract_stat1_peaks.fa</em><br><br>
            Total Sequence length: 4863558<br>
            0.24,0.26,0.26,0.24<br><br>

        <h3> How to build Bowtie index files</h3>
        <p class="tab">
          Let's define BOWTIE_DIR as as the directory containing the bowtie indices and ASSEMBLY_DIR as the one where the FASTA chromosome files for a specific assembly (e.g. hg19) are stored.<br>
          Go to the BOWTIE_DIR directory, and concatenate all chromosome files into a sequence file <em style="color:blue">h_sapiens_hg19.fa</em> that includes the entire hg19 genome assembly:
          <br><br>
            ls -1 -v ASSEMBLY_DIR/chrom*.seq | xargs cat &gt; h_sapiens_hg19.fa
        </p>
        <p class="code">
          <span style="color:firebrick"># Example:</span><br>
          <em> cd ../genomedb/bowtie</em><br>
          <em> ls -1 -v ../hg19/chrom*.seq | xargs cat &gt; h_sapiens_hg19.fa</em><br><br>
        </p>
        <p class="tab">
          You then execute the following command:
        </p>
        <p class="code">
          bowtie-build -f h_sapiens_hg19.fa h_sapiens_hg19
        </p>
    </body>
</html>
