<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
        "http://www.w3.org/TR/html4/loose.dtd">
<html>
    <head>
        <title>README - Examples</TITLE>
        <style type="text/css">
            .tab { margin-left: 40px; }
            .code { color: Teal ; 
                    font-size: 100%; 
                    font-style: italic; 
                    line-height: 1.5;
                    padding-left: 20px;
                    border-style: solid;
                    margin-left: 20px;
                    margin-right: 200px;
            }
            .s { color: Teal ; 
                    font-size: 100%; 
                    font-style: italic; 
                    line-height: 1.5;
                    padding-left: 20px;
                    border-style: solid;
                    margin-left: 40px;
                    margin-right: 200px;
            }
        </style>
    </head>
    
    <body style='background-color:#F2F2F2; font-size:14px; text-align: justify; width:90%'>
                
        <h1>Application Example</h1> 

        <h2>Scan the hg19 human genome with the chen10 CTCF Position Weight Matrix (PWM)</h2>

        <h3> Important remarks: </h3>
        <p class="tab">
          The PWMScan programs rely on a few rules to make output conversion to BED format easier, namely FASTA header and file naming conventions.
        <p class="tab">
        
          The final output format is <b>BEDdetail</b>, which is an extension of BED format that is used to enhance the track display page.
          For PWMScan, we use BEDdetail format to include the name of the PWM as well as the p-value associated to the sites identified by PWMScan.

        <p class="tab">
          The <b>BEDdetail</b> format provides the following obligatory fields:
        <ul style="list-style: none;">
          <li>1) Chromosome name (e.g. chr1, chrX, etc.)</li>
          <li>2) Starting position of the match</li>
          <li>3) Ending position of the match</li>
          <li>4) Matching nucleotide sequence</li>
          <li>5) Integer score of PWM match</li>
          <li>6) Strand (+, -)</li>
          <li>7) PWM name</li>
          <li>8) P-value of the match</li>
        </ul>
         <p class="tab">

        <h4> 
            <p class="tab">
            FASTA headers  
        </h4>
        
        <p class="tab">
        All genome sequence files have a FASTA header that is formatted as follows:
        <pre>
            >chr|NC_000001|NC_000001.10 <i>some text</i>
        </pre>
        
        <p class="tab">
        The <i><font color=blue>sequence identifier</font></i> includes three words concatenated via a UNIX pipe '|': the word 'chr' followed by the NCBI RefSeq partial and full accession identifiers. 
        The accession records indicate sequence identity. The sequence identifier is parsed by most of the PWMScan programs and utilities.
        <p class="tab">
        For each genome assembly, we provide a table for NCBI RefSeq identifier to chromosome number conversion (<i><font color=blue>chr_NC_gi</font></i>) as well as a table for FASTA sequence identifier to chromosome 
        number conversion (<i><font color=blue>chr_hdr</font></i>). The <i><font color=blue>chr_NC_gi</font></i> file has been downloaded from NCBI whereas the <i><font color=blue>chr_hdr</font></i> file has been locally-generated. By consequence, the <i><font color=blue>chr_hdr</font></i> table can be changed to adapt to other FASTA header format conventions.
        <p class="tab">
        Here is an example for the human genome assembly hg19:
        <pre>
            <i><font color=firebrick>chr_NC_gi</font></i>
            #Chr    Accession.ver   gi
            1       NC_000001.10    224384768
            2       NC_000002.11    224384767
            3       NC_000003.11    224384766
            4       NC_000004.11    224384765
            5       NC_000005.9     224384764
            6       NC_000006.11    224384763
            7       NC_000007.13    224384762
            8       NC_000008.10    224384761
            9       NC_000009.11    224384760
            10      NC_000010.10    224384759
            11      NC_000011.9     224384758
            12      NC_000012.11    224384757
            13      NC_000013.10    224384756
            14      NC_000014.8     224384755
            15      NC_000015.9     224384754
            16      NC_000016.9     224384753
            17      NC_000017.10    224384752
            18      NC_000018.9     224384751
            19      NC_000019.9     224384750
            20      NC_000020.10    224384749
            21      NC_000021.8     224384748
            22      NC_000022.10    224384747
            X       NC_000023.10    224384746
            Y       NC_000024.9     224384745
            M       NC_012920.1     251831106
        </pre>
        <pre>
            <i><font color=firebrick>chr_hdr</font></i>
            #Chr    Sequence Header                 Assembly
            1       chr|NC_000001|NC_000001.10      hg19
            2       chr|NC_000002|NC_000002.11      hg19
            3       chr|NC_000003|NC_000003.11      hg19
            4       chr|NC_000004|NC_000004.11      hg19
            5       chr|NC_000005|NC_000005.9       hg19
            6       chr|NC_000006|NC_000006.11      hg19
            7       chr|NC_000007|NC_000007.13      hg19
            8       chr|NC_000008|NC_000008.10      hg19
            9       chr|NC_000009|NC_000009.11      hg19
            10      chr|NC_000010|NC_000010.10      hg19
            11      chr|NC_000011|NC_000011.9       hg19
            12      chr|NC_000012|NC_000012.11      hg19
            13      chr|NC_000013|NC_000013.10      hg19
            14      chr|NC_000014|NC_000014.8       hg19
            15      chr|NC_000015|NC_000015.9       hg19
            16      chr|NC_000016|NC_000016.9       hg19
            17      chr|NC_000017|NC_000017.10      hg19
            18      chr|NC_000018|NC_000018.9       hg19
            19      chr|NC_000019|NC_000019.9       hg19
            20      chr|NC_000020|NC_000020.10      hg19
            21      chr|NC_000021|NC_000021.8       hg19
            22      chr|NC_000022|NC_000022.10      hg19
            X       chr|NC_000023|NC_000023.10      hg19
            Y       chr|NC_000024|NC_000024.9       hg19
            M       chr|NC_012920|NC_012920.1       hg19
        </pre>
        
        <h4> 
            <p class="tab">
                Files and paths
            </p>
        </h4>
        
        <p class="tab">
            The tar files <i><font color=blue>genomes_chr_NC_gi.tar.gz</font></i> and <i><font color=blue>genomes_chr_hdr.tar.gz</font></i> include a list of <i><font color=firebrick>chr_NC_gi</font></i> and <i><font color=firebrick>chr_hdr</font></i> tables for several common model organisms such as human, mouse, fruit fly, worm, zebrafish, yeast, and more. As mentioned above, this tables are used to convert match lists to BED format.
        </p>
        
        <p class="tab">
            To check the content, type:
        
        <p class="s">
            tar -tvzf genomes_chr_NC_gi.tar.gz<br>
            tar -tvzf genomes_chr_hdr.tar.gz
        </p>
    
        <p class="tab">
            To extract the files, choose a convenient directory (<font color=firebrick>genome_root_dir</font>), and execute
            the following commands:
            
        <p class="s">
            cp genomes_chr_NC_gi.tar.gz genomes_chr_hdr.tar.gz genome_root_dir/<br>
            cd genome_root_dir<br>
            tar -xvzf genomes_chr_NC_gi.tar.gz<br>
            tar -xvzf genomes_chr_hdr.tar.gz
        </p>
        
        <p class="tab">
            The default location (variable genome_root_dir) for conversion programs and scripts is the following:<br>
        <p class="tab">
            <font color=blue>genome_root_dir=<i>/home/local/db/genome</i></font><br>
        <p class="tab">
            which corresponds to the root directory of the entire genome assembly data.<br>
            This can be changed via the command option -d  <font color=firebrick>&lt;genome-root-dir&gt;</font> in all bash wrapper scripts (as described below).<br>
            The PWM for the CTCF transcription factor that will be used in our example is the <font color=blue><i>chen10_ctcf.mat</i></font> file.
        </p>
        
        
        
        <h3>Analysis Pipeline using two different scanning methods</h3>
        <ul style='list-style-type: none;'>
        <li><b>1) Calculate the matrix score corresponding to p-value=0.00001 (10<sup>-5</sup>)</b></li>
        
        <p class="code">
            matrix_prob -e 0.00001 --bg "0.29,0.21,0.21,0.29" chen10_ctcf.mat
        </p>
            
        <pre>
           SCORE :   1128  PERC :  96.90%
        </pre>
        <p>
           There are two basic methods to scan the genome with a PWM and a cut-off/p-value:
           <ul style='list-style-type: none;'>
               <li>1- Use a fast string-matching algorithm such us <font color=firebrick>Bowtie</font> as follows:
               <ul style='list-style-type: circle;'> 
               <li>Given a PWM and a cut-off or p-value, generate all possible matches that represent the PWM with a score greater or equal to the cut-off;
               <li>Map the list of all possible PWM matches to a reference genome ar a set of DNA sequences of interest using Bowtie.
                   (<font color=firebrick>Bowtie</font>).
               </ul>
               <li>2- Use <font color=firebrick>matrix_scan</font>, a C program using a conventional search algorithm that has been optimized for rapid score computation and drop-off strategy.
            </ul>
           <p> 
           If you are interested in using the first approach, you should read sections 3) and 4), whereas for using matrix_scan please refer to section 5).
        <p>    
            <li>
                <b>2) Generate the PWM Score Cumulative Distribution</b><p>
                <font color=black>The cumulative score distribution is used to assign a p-value to a match score.
                The matrix_prob program is used to compute the cumulative score distribution.</font>
            </li>
            
            <p class="code">
                matrix_prob --bg "0.29,0.21,0.21,0.29" chen10_ctcf.mat &gt; chen10_ctcf_co1128_scoretab.txt
            </p>

            <li>
                <b>3) Generate a list of all tags that represent all possible PWM matches (with minimal score of 1128)</b><p>
                <font color=firebrick>This step is only necessary if fast string-matchig tools, such as Bowtie,
                are used to scan the genome.</font>
            </li>
            
            <p class="code">
                mba -c 1128 -l 19 chen10_ctcf.mat &gt; chen10_ctcf_co1128.dat<br>
                <font color=firebrick># Convert the tag list into FASTA format (for Bowtie):</font><br>
                awk '{print "&gt;"$2"\n"$1}' chen10_ctcf_co1128.dat &gt; chen10_ctcf_co1128_taglist.fa<br>
            </p>

            The reason for generating a match list in FASTA format is that we want to carry the match score over to the next steps (see below).
           <p>

            <li>
                <b>4) Map the tags representing the PWM to hg19</b>
           <p>
                PWMScan uses the program Bowtie to map all matching tags to the hg19 genome assembly.
            </li>
        
           <p> 
                <li>
                    <b><font color=firebrick>1- Bowtie</font></b>
                    <p>
                       Let's define GENOME_DIR as the genome root directory:
                        <ul style='list-style-type: none;'>
                            <li><font color=blue>GENOME_DIR = /home/local/db/genome</font></li>
                            <li><font color=blue>chr_NC_PATH = GENOME_DIR</font></li>
                        </ul>
                    <p>
                    <p>
                        Let's define BOWTIE_DIR as the directory containing the Bowtie indexed genomes:
                        <ul style='list-style-type: none;'>
                            <li><font color=blue>h_sapiens_hg19</font> identifies the index files linked to the hg19 human assembly</li>
                            <li><font color=blue>BOWTIE_DIR = GENOME_DIR/bowtie</font></li>
                        </ul>
                    <p>
                        The command to map the list of PWM tags to hg19 is the follwing:
                    <p>    
                        <i>bowtie -l 19 -n0 -a BOWTIE_DIR/h_sapiens_hg19 -f chen10_ctcf_co1128_taglist.fa --un unmapped.dat &gt; chen10_ctcf_co1128_bowtie.out</i>
                    <p>    
		        It takes of the order of 12 seconds for a full scan of the hg19 genome assembly (with the exclusion of the mitochondrial chromosome). The total number of hits is 143597.</font><br>
                    
                    <p class="code">
                            <font color=firebrick># Example:</font>
                            <br>                        
                            bowtie --threads 4 -l 19 -n0 -a /home/local/db/genome/bowtie/h_sapiens_hg19 -f chen10_ctcf_co1128_taglist.fa --un unmapped.dat &gt;chen10_ctcf_co1128_bowtie.out<br>
                            <font color=firebrick># Convert the bowtie output to <b><span style='font-size:13px'>BEDdetail</span></b> via the following command:</font><br>
                            sort -s -k3,3 -k4,4n chen10_ctcf_co1128_bowtie.out | bowtie2bed -s hg19 -l 19 -i chr_NC_PATH | awk 'BEGIN { while((getline line &lt; "chen10_ctcf_co1128_scoretab.txt") &gt; 0 ) {split(line,f," "); pvalue[f[1]]=f[2]} close("chen10_ctcf_co1128_scoretab.txt")} {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t""chen10_ctcf""\t""P-value="pvalue[$5]}' &gt; chen10_ctcf_co1128_bowtie.bed<br>

                            <font color=firebrick># Example (use default chr_NC_PATH):</font><br>
                            sort -s -k3,3 -k4,4n chen10_ctcf_co1128_bowtie.out | bowtie2bed -s hg19 -l 19 -i /home/local/db/genome | awk 'BEGIN { while((getline line &lt; "chen10_ctcf_co1128_scoretab.txt") &gt; 0 ) {split(line,f," "); pvalue[f[1]]=f[2]} close("chen10_ctcf_co1128_scoretab.txt")} {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t""chen10_ctcf""\t""P-value="pvalue[$5]}' &gt; chen10_ctcf_co1128_bowtie.bed<br>
                            <br>
		            <font color=firebrick># Note: to convert Bowtie output to BED format, we use the chr_hdr file for assembly hg19.<br>
                            # By default, the chr_hdr file is located at chr_NC_PATH/hg19 = /home/local/db/genome/hg19.<br>
                            # chr_NC_PATH can be changed by using option -i.<br>
                            # Also note that the bowtie ouput reports the match score in the first field so that it can be easily retrieved.<br>
                            <br>
                            # To run the bowtie-based pipeline, type:</font><br>
                            bowtie --threads 4 -l 19 -n0 -a BOWTIE_DIR/h_sapiens_hg19 -f chen10_ctcf_co1128_taglist.fa --un unmapped.dat | sort -s -k3,3 -k4,4n | 
                            bowtie2bed -s hg19 -l 19 -i chr_NC_PATH | awk 'BEGIN { while((getline line &lt; "chen10_ctcf_co1128_scoretab.txt") &gt; 0 ) {split(line,f," "); pvalue[f[1]]=f[2]} close("chen10_ctcf_co1128_scoretab.txt")} {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t""chen10_ctcf""\t""P-value="pvalue[$5]}' &gt; chen10_ctcf_co1128_bowtie.bed<br>
                            <font color=firebrick># Bowtie can read input files from stdin. You should specify "-" for stdin.<br>
                            # In such case, you can run the entire pipeline as follows:</font><br>
                            mba -c 1128 -l 19 chen10_ctcf.mat | awk '{print "&gt;"$2"\n"$1}' | bowtie --threads 4 -l 19 -n0 -a BOWTIE_DIR/h_sapiens_hg19 -f - --un unmapped.dat | sort -s -k3,3 -k4,4n | bowtie2bed -s hg19 -l 19 -i chr_NC_PATH | awk 'BEGIN { while((getline line &lt; "chen10_ctcf_co1128_scoretab.txt") &gt; 0 ) {split(line,f," "); pvalue[f[1]]=f[2]} close("chen10_ctcf_co1128_scoretab.txt")} {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t""chen10_ctcf""\t""P-value="pvalue[$5]}' &gt; chen10_ctcf_co1128_bowtie.bed<br>
                            <font color=firebrick># Example:</font><br>
                             mba -c 1128 -l 19 chen10_ctcf.mat | awk '{print "&gt;"$2"\n"$1}' | bowtie --threads 4 -l 19 -n0 -a /home/local/db/genome/bowtie/h_sapiens_hg19 -f - --un unmapped.dat | sort -s -k3,3 -k4,4n | bowtie2bed -s hg19 -l 19 | awk 'BEGIN { while((getline line &lt; "chen10_ctcf_co1128_scoretab.txt") &gt; 0 ) {split(line,f," "); pvalue[f[1]]=f[2]} close("chen10_ctcf_co1128_scoretab.txt")} {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t""chen10_ctcf""\t""P-value="pvalue[$5]}' &gt; chen10_ctcf_co1128_bowtie.bed<br>                           
                        </p>
                </li>
        <p>
        <li><b>5) Using matrix_scan</b> 
        <p>
           Let's define GENOME_DIR as the genome root directory:
        <p>
             <ul style='list-style-type: none;'>
             <li><font color=blue>GENOME_DIR = /home/local/db/genome</font>
             <li><font color=blue>chr_NC_PATH = GENOME_DIR</font>
             </ul>
        <p>    
          We define as ASSEMBLY_DIR the sub-directory containing all the hg19 chromosome FASTA files:
        <p>
             <ul style='list-style-type: none;'>
             <li><font color=blue>ASSEMBLY_DIR = GENOME_DIR/hg19</font>
             </ul>
        <p>    
           The command to scan the hg19 genome against the chen10 CTCF PWM is the follwing:
        <p>
           <i>cat ASSEMBLY_DIR/chrom*.seq | matrix_scan -m chen10_ctcf.mat -c 1128 | sort -s -k1,1 -k2,2n -k6,6 > chen10_ctcf_co1128_matrix_scan.out</i>
        <p>
           It takes of the order of 3 minutes to scan the entire genome. The total number of hits is 143597 (if we exclude the mitochondrial chromosome) in agreement with the tag mapping methods.<br>
           <font color=firebrick>Note that matrix_scan reports the match score in the fifth field of its output.</font><br>
           <font color=firebrick>Also note that <i>matrix_scan</i> parses the sequence FASTA header assuming that the sequence identifier is of the following type:</font><br>
        <pre>
        word|Accession|Accession
        Ex:
        chr|NC_000004|NC_000004.11 
        </pre>
        <p>
           <font color=firebrick># Parallelization of matrix_scan</font><br>
           <font color=black>To improve performance, we can run matrix_scan in parallel (on each chromosome file) using a simple python script called <b>matrix_scan_parallel.py</b> that dispatches the jobs to available CPU-cores. In this way, matrix_scan competes with the Bowtie-based approach.</font>
        <p>
           <p class="code">
           <font color=firebrick># Example:<br></font>
           <i>cat /home/local/db/genome/hg19/chrom*.seq | matrix_scan -m chen10_ctcf.mat -c 1128 | sort -s -k1,1 -k2,2n -k6,6 &gt; chen10_ctcf_co1128_matrix_scan.out</i>
        <br>
           <font color=firebrick># To use matrix_scan_parallel.py, type the following command:<br> </font>
           <i>python /local/bin/matrix_scan_parallel.py -m chen10_ctcf.mat -f "$(ls /home/local/db/genome/hg19/chrom*.seq|grep -v chromMt)" -c 1128 -p 15 | sort -s -k1,1 -k2,2n -k6,6 &gt; chen10_ctcf_co1128_matrix_scan.out</i>
        <br><br>
           <font color=firebrick># The -p option is used to set the maximum number of parallel processes for execution of matrix_scan. The above example takes only about 20 seconds to complete the task in agreement with Bowtie.<br><br>
           # Convert the matrix scan output to <b><span style='font-size:13px'>BEDdetail</span></b> by issuing the following command:</font>
        <br>
           <i>mscan2bed -s hg19 -i chr_NC_PATH chen10_ctcf_co1128_matrix_scan.out | awk 'BEGIN { while((getline line &lt; "chen10_ctcf_co1128_scoretab.txt") &gt; 0 ) {split(line,f," "); pvalue[f[1]]=f[2]} close("chen10_ctcf_co1128_scoretab.txt")} {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t""chen10_ctcf""\t""P-value="pvalue[$5]}' &gt; chen10_ctcf_co1128_matrix_scan.bed</i>
        <br>
           <font color=firebrick># Example:<br></font>
           <i>mscan2bed -s hg19 -i /home/local/db/genome chen10_ctcf_co1128_matrix_scan.out | awk 'BEGIN { while((getline line &lt; "chen10_ctcf_co1128_scoretab.txt") &gt; 0 ) {split(line,f," "); pvalue[f[1]]=f[2]} close("chen10_ctcf_co1128_scoretab.txt")} {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t""chen10_ctcf""\t""P-value="pvalue[$5]}' &gt; chen10_ctcf_co1128_matrix_scan.bed</i>
        <br><br>
           <font color=firebrick>
           # Note : to convert matrix_scan output to BED format, we use the chr_NC_gi file for assembly hg19.<br>
           # By default, the chr_NC_gi is located at chr_NC_PATH/hg19 = /home/local/db/genome/hg19.<br> 
             &nbsp;&nbsp;chr_NC_PATH can be changed by using the &lt;-i&gt; option. 
        <br><br>   
           # To run the entire pipeline:
        <br></font>
           <i>cat ASSEMBLY_DIR/chrom*.seq | matrix_scan -m chen10_ctcf.mat -c 1128 | sort -s -k1,1 -k2,2n -k6,6 | mscan2bed -s hg19 -i chr_NC_PATH | awk 'BEGIN { while((getline line &lt; "chen10_ctcf_co1128_scoretab.txt") &gt; 0 ) {split(line,f," "); pvalue[f[1]]=f[2]} close("chen10_ctcf_co1128_scoretab.txt")} {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t""chen10_ctcf""\t""P-value="pvalue[$5]}' &gt; chen10_ctcf_co1128_matrix_scan.bed</i>
        <br>
           <font color=firebrick># Example:</font>
        <br>
           <i>cat /home/local/db/genome/hg19/chrom*.seq | matrix_scan -m chen10_ctcf.mat -c 1128 | sort -s -k1,1 -k2,2n -k6,6 | mscan2bed -s hg19 -i /home/local/db/genome | awk 'BEGIN { while((getline line &lt; "chen10_ctcf_co1128_scoretab.txt") &gt; 0 ) {split(line,f," "); pvalue[f[1]]=f[2]} close("chen10_ctcf_co1128_scoretab.txt")} {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t""chen10_ctcf""\t""P-value="pvalue[$5]}' &gt; chen10_ctcf_co1128_matrix_scan.bed</i>
        <br>
           <font color=firebrick># Using parallelization:</font>
        <br>
           <i>python /local/bin/matrix_scan_parallel.py -m chen10_ctcf.mat -f "$(ls /home/local/db/genome/hg19/chrom*.seq|grep -v chromMt)" -c 1128 -p 15 | sort -s -k1,1 -k2,2n -k6,6 | mscan2bed -s hg19 -i /home/local/db/genome | awk 'BEGIN { while((getline line &lt; "chen10_ctcf_co1128_scoretab.txt") &gt; 0 ) {split(line,f," "); pvalue[f[1]]=f[2]} close("chen10_ctcf_co1128_scoretab.txt")} {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t""chen10_ctcf""\t""P-value="pvalue[$5]}' &gt; chen10_ctcf_co1128_matrix_scan.bed</i>
        <p>
        </p>
           </li></ul>
        
        <p>
        <h3> Shell (bash) Wrappers to execute the analysis pipeline</h3>
         <p class="tab">
         These shell wrapper scripts have been written to make it easier to run the whole pwmscan pipeline, irrespecitve of what method one chooses.
        <br>
         They require to specify the matrix file containing the integer PWM, the p-value, the path to the assembly or index files, and the UCSC assembly name for the species (e.g. hg19, hg38, mm9 etc.).
         Optional parameters are the overlapping matches flag (-o), the forward scanning (-f) option, the parallel execution (-p) and write to file (-w) flags.
        <br>
        If the '-o' option is specified, overlapping matches are retained. The '-f' option activates forward scanning. If the '-p' option is set, the matrix_scan program execution is done by parallel dispatch. If the -w option is set, the results are written to file, else output goes to STDOUT.
        <br>
        The output file is a match list in <b>BEDdetail</b> format.
        <br>
        <br>
        The <b>pwm_scan</b> script scans a genome with a PWM using either Bowtie or matrix_scan depending on both the p-value and the matrix length.
        For high p-values and long motifs, the Bowtie-based strategy becomes inefficient and the matrix_scan-based approach is the fastest one.
        <br>
        The <b>pwm_bowtie_wrapper</b> script implements the Bowtie-based pipeline whereas the <b>pwm_mscan_wrapper</b> script uses the matrix_scan-based approach.
        <br>
        <br>
        The <b>pwmlib_scan</b> script scans a genome with a collection of PWMs coming from a database.<br>
        Accepted motif database formats are the MEME libraries as well as the motif libraries (both log-odds and letter-probability formats) provided by the PWMScan Web interface.
 
         <p class="tab">
         <b>NOTE:</b> The path to all the PWMScan executables and scripts is defined by the <b>bin_dir</b> variable and is hard-coded in all bash scripts, and is changed upon installation via the Makefile, so that all installed scripts in <b>bin_dir</b> have the correct binary pathname.

        <p class="tab" style='font-weight: bold; font-size:13px'>
         A few remarks on the general <i>pwm_scan script</i>.
        <p class="tab">
         The &lt;genome-root-dir&gt; input argument is the root directory of the genome files (the Bowtie index files, and the FASTA chromosome files used by matrix_scan, and the assembly tables used to convert the PWMScan pipeline output to BED format).
        <br>
         &lt;genome-root-dir&gt; is supposed to have two sub-directories types, the /bowtie sub-directory where the Bowtie index files are stored, and the /&lt;assembly&gt; sub-directory for storing assembly-specific chromosome files.
        <p class="tab">
         In our setting, <b>&lt;genome-root-dir&gt;</b> is set to <span style='font-weight: bold; font-size:13px'><i>/home/local/db/genome</i></span>.
        <br>
         The <b>chrNC_dir</b> variable (used for locating the the chr_NC_gi/chr_hdr files) is set to chrNC_dir=<b>&lt;genome-root-dir&gt;</b>.
        <br>
         <ul style='list-style-type: none;'>
         <li><font color=firebrick><b>1) Bowtie (pwm_bowtie_wrapper)</b> script</font>
        <p class="code">
           # Usage: <br>pwm_bowtie_wrapper -m &lt;matrix-file&gt; -e &lt;p-value&gt; -d &lt;genome-root-dir&gt; -s &lt;assembly[hg19|mm9|..]&gt; [options]<br>
        - version 1.1.4<br>
        where options are:<br>
             -f  &nbsp;&nbsp;Scan sequences in forward direction [def=bidirectional]<br>
             -o  &nbsp;&nbsp;Allow for overlapping matches [def=non-overlapping matches]<br>
             -w  &nbsp;&nbsp;Write output to file [def=STDOUT]<br>
             -p  &nbsp;&nbsp;Distribute matrix_scan on multiple CPU-cores [def=non-parallel processing]<br>
        <br>
           <font color=firebrick># Example:</font>
        <br>
           <i>pwm_bowtie_wrapper -m chen10_ctcf.mat -e 0.00001 -d /home/local/db/genome -s hg19 -o -w</i>
        <br>
           <font color=firebrick># List of matches (BED format) : chen10_ctcf_co1128_bowtie.bed</font>
        <br>
           <font color=firebrick># Total number of PWM matches  : 143597</font>
        <br>
        </p>

        </li><li><font color=firebrick><b>2) Matrix Scan (pwm_mscan_wrapper)</b> script</font>
        <p class="code">
           # Usage:<br> pwm_mscan_wrapper -m &lt;matrix-file&gt; -e &lt;p-value&gt; -d &lt;genome-root-dir&gt; -s &lt;assembly[hg19|mm9|..]&gt; [options]<br>
        - version 1.1.5<br>
        where options are:<br>
             -f  &nbsp;&nbsp;Scan sequences in forward direction [def=bidirectional]<br>
             -o  &nbsp;&nbsp;Allow for overlapping matches [def=non-overlapping matches]<br>
             -w  &nbsp;&nbsp;Write output to file [def=STDOUT]<br>
             -p  &nbsp;&nbsp;Distribute matrix_scan on multiple CPU-cores [def=non-parallel processing]<br>
        <br>
           <font color=firebrick># Example:</font>
        <br>
             <i>pwm_mscan_wrapper -m chen10_ctcf.mat -e 0.00001 -d /home/local/db/genome -s hg19 -o -p -w</i>
        <br>
           <font color=firebrick># List of matches (BED format) : chen10_ctcf_co1128_matrix_scan.bed</font>
        <br>
           <font color=firebrick># Total number of PWM matches  : 143597</font>
       <br>
        </p>

        </li><li><font color=firebrick><b>3) PWM Scan (pwm_scan)</b> script</font>
        <p class="code">
           # Usage:<br> pwm_scan -m &lt;matrix-file&gt; -e &lt;p-value&gt; -d &lt;genome-root-dir&gt; -s &lt;assembly[hg19|mm9|..]&gt; [options]<br>
        - version 1.1.5<br>
        where options are:<br>
             -f  &nbsp;&nbsp;Scan sequences in forward direction [def=bidirectional]<br>
             -o  &nbsp;&nbsp;Allow for overlapping matches [def=non-overlapping matches]<br>
             -w  &nbsp;&nbsp;Write output to file [def=STDOUT]<br>
             -p  &nbsp;&nbsp;Distribute matrix_scan on multiple CPU-cores [def=non-parallel processing]<br>
        <br>
           <font color=firebrick># Examples:</font>
        <br>
             <i>pwm_scan -m chen10_ctcf.mat -e 0.00001 -d /home/local/db/genome -s hg19 -o -w</i>
        <br>
           <font color=firebrick># List of matches (BED format) : chen10_ctcf_co1128_bowtie.bed</font>
        <br>
           <font color=firebrick># Or if we set the &lt;parallel&gt; option:</font>
        <br>
             <i>pwm_scan -m chen10_ctcf.mat -e 0.00001 -d /home/local/db/genome -s hg19 -o -w -p</i>
        <br>
           <font color=firebrick># List of matches (BED format) : chen10_ctcf_co1128_matrix_scan.bed</font>
        <br>
           <font color=firebrick># Also try the following commands (with p-value=10<sup>-6</sup>):</font>
        <br>
             <i>pwm_scan -m chen10_ctcf.mat -e 0.000001 -d /home/local/db/genome -s hg19 -w -p</i>
        <br>
           <font color=firebrick># List of matches (BED format) : chen10_ctcf_co1549_bowtie.bed</font>
        <br>
             <i>pwm_scan -m stat1.mat -e 0.000001 -d /home/local/db/genome -s hg19 -o -w -p</i>
        <br>
           <font color=firebrick># List of matches (BED format) : stat1_co1731_bowtie.bed</font>
        <br>
           <font color=firebrick># N.B.</font>
        <br>
           <font color=black> If the &lt;parallel&gt; option is set (default), matrix_scan is run in parallel by splitting the processing of the entire genome on multiple processes for each chromosome file in parallel.
        <br>
           Of course, this only works on a multi-core processor machine.
        <br>
        </p>

        </li><li><font color=firebrick><b>4) PWMlib Scan (pwmlib_scan)</b> script</font>
        <p class="code">
           # Usage:<br> pwmlib_scan -l &lt;pwmlib-file&gt; -e &lt;p-value&gt; -d &lt;genome-root-dir&gt; -s &lt;assembly[hg19|mm9|..]&gt; [options]<br>
        - version 1.1.5<br>
        where options are:<br>
             -g  &nbsp;&nbsp;PWM format. Accepted formats are : meme, logodds [def], and lpm<br>
             -f  &nbsp;&nbsp;Scan sequences in forward direction [def=bidirectional]<br>
             -o  &nbsp;&nbsp;Allow for overlapping matches [def=non-overlapping matches]<br>
             -w  &nbsp;&nbsp;Write output to file [def=STDOUT]<br>
             -p  &nbsp;&nbsp;Distribute matrix_scan on multiple CPU-cores [def=non-parallel processing]<br>
             -v  &nbsp;&nbsp;Verbose mode and keep PWM tmp files<br>
        <br>
           <font color=firebrick># Examples:</font>
        <br>
        <br>
           <font color=firebrick># Scan the human genome hg19 with the JASPAR CORE 2018 vertebrates collection (meme format):</font>
        <br>
             <i>pwmlib_scan -l JASPAR_CORE_2018_vertebrates.meme -g meme -e 0.0001 -d /home/local/db/genome -s hg19 -w -p 2&gt;pwmlib_scan_log.txt &</i>
        <br>
           <font color=firebrick># Scan the human genome hg19 with the JASPAR CORE 2018 vertebrates collection (lpm format):</font>
        <br>
             <i>pwmlib_scan -l JASPAR_CORE_2018_vert_matrix_probs.mat -g lpm -e 0.0001 -d /home/local/db/genome -s hg19 -w -p 2&gt;pwmlib_scan_log.txt &</i>
        <br>
          </li></ul>

        <h3> Shell (bash) Wrapper for matrix conversion</h3>
         <p class="tab">
         The pwm_convert bash script is used to convert JASPAR, TRANSFAC, PFM, LPM and real PWM formats to integer log-odds format.
         <ul style='list-style-type: none;'>
        <p class="code">
           # Usage: pwm_convert &lt;matrix-file&gt; -f=&lt;format [jaspar|transfac|lpm|pfm|real]&gt;
             <br>
 &nbsp;&nbsp; OPTIONS:
             <br>
&nbsp;&nbsp;&nbsp;&nbsp;  -h                  show help message 
             <br>
&nbsp;&nbsp;&nbsp;&nbsp;  -b=<bg-freq-file>   bg nucleotide composition file 
             <br>
&nbsp;&nbsp;&nbsp;&nbsp;  -c=<pseudo-cnt>     pseudo-count fraction [def=0] 
             <br>
&nbsp;&nbsp;&nbsp;&nbsp;  -m=<minscore>       minimal score value [def=-10000] 
             <br>
&nbsp;&nbsp;&nbsp;&nbsp;  -n=<log-scaling>    log scaling factor [def=100] 
             <br>
&nbsp;&nbsp;&nbsp;&nbsp;  -s=<mult>           matrix multiplication factor (for real PWMs) [def=100] 
             <br>
&nbsp;&nbsp;&nbsp;&nbsp;  Convert JASPAR, TRANSFAC, PFM, LPM and real PWM formats to integer log likelihoods (log-odds) format.
        <br>
         <font color=black>By default, the prior background nucleotide frequencies are set to 0.25.</font><br>

        <p class="code">
        <font color=firebrick># Examples</font>
        <br>
        <font color=firebrick># Convert JASPAR to log-odds:</font>
        <br>
             <i>pwm_convert stat1_jaspar.mat -f=jaspar 2&gt;/dev/null</i><br>

        <font color=firebrick># If we want to use different backgroung nucleotide frequencies:</font>
        <br>
             <i>pwm_convert stat1_jaspar.mat -f=jaspar -b=bg_probs_2.txt 2&gt;/dev/null</i><br>

        <font color=black># Where the bg_probs_2.txt file includes the prior bg frequencies, defined as follows:</font>
        <br>
         &nbsp;&nbsp;A  0.29 <br>
         &nbsp;&nbsp;C  0.21 <br>
         &nbsp;&nbsp;G  0.21 <br>
         &nbsp;&nbsp;T  0.29 <br>

        <font color=firebrick># Convert TRANSFAC to log-odds:</font>
        <br>
             <i>pwm_convert statx_transfac.mat -f=transfac -c=0.0000001 2&gt;/dev/null</i><br>

        <font color=firebrick># Convert PFM (Position Frequency Matrix) to log-odds:</font>
        <br>
             <i>pwm_convert pwm_convert stat1_pfm.mat -f=pfm 2&gt;/dev/null</i><br>
        <br>
        </ul>
        <p>
        <h3> How to build Bowtie index files</h3>
        <p class="tab">
         <ul style='list-style-type: none;'>
        <p>
          Let's define BOWTIE_DIR as as the directory containing the bowtie indices and ASSEMBLY_DIR as the one where the FASTA chromosome files for a specific assembly (e.g. hg19) are stored.<br>
          Go to the BOWTIE_DIR directory, and concatenate all chromosome files into a sequence file that includes the entire hg19 genome assembly:
        <p class="code">
          ls -1 -v ASSEMBLY_DIR/chrom*.seq | xargs cat &gt; h_sapiens_hg19.fa
        </p>
          You then execute the following command:
        <p class="code">
          bowtie-build -f h_sapiens_hg19.fa h_sapiens_hg19
        </p>

    </body>
</html>
